kedro/pipeline/node.py

```diff
@@ -397,38 +399,55 @@ def _run_with_dict(self, inputs: Dict[str, Any], node_inputs: Dict[str, str]):
 
     def _outputs_to_dictionary(self, outputs):
         def _from_dict():
-            if set(self._outputs.keys()) != set(outputs.keys()):
+            (result,), iterator = (
+                ((outputs,), None)
+                if not inspect.isgenerator(outputs)
+                else spy(outputs, 1)
+            )
+            keys = list(self._outputs.keys())
+            if not isinstance(result, dict):
+                raise ValueError(
+                    f"Failed to save outputs of node {self}.\n"
+                    f"The node output is a dictionary, whereas the "
+                    f"function output is not."
+                )
+            if set(keys) != set(result.keys()):
                 raise ValueError(
                     f"Failed to save outputs of node {str(self)}.\n"
-                    f"The node's output keys {set(outputs.keys())} do not match with "
-                    f"the returned output's keys {set(self._outputs.keys())}."
+                    f"The node's output keys {set(result.keys())} "
+                    f"do not match with the returned output's keys {set(keys)}."
                 )
-            return {name: outputs[key] for key, name in self._outputs.items()}
+            if iterator:
+                exploded = map(lambda x: tuple(x[k] for k in keys), iterator)
```

Comment:
The scope for this variable is rather short and it's not untypical for short lambda functions to use just `x`, so I am following that convention.

But also yes, it'd be rather difficult to know what exactly `x` is, in general it will be a dictionary object, returned by each separate `yield` call in the node function.